src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java

```diff
@@ -158,9 +158,49 @@ private class ResolvedIndicesProvider implements IndicesProvider {
         private final ImmutableSet.Builder<String> remoteIndices;
         // set of previously resolved index requests to avoid resolving
         // the same index more than once while processing bulk requests
-        private final Set<MultiKey> alreadyResolved;
+        private final Set<AlreadyResolvedKey> alreadyResolved;
```

Comment:
Not sure about it. But using unbounded collection could lead to a memory leak. So a question to those who knows this part better than me. Am I right or not?

Reply:
If we wanted to prevent the chance of a memory leak entirely, I believe we would need to cap the size of the set. The problem is determining an appropriate cap for the size. I believe the set should be cleared with each request so it should not persist for too long, so the issue would only be that we ran into a memory issue in a single bulk request. For determining the size to cap it at, we could look at the expected size of each AlreadyResolvedKey object and then determine how much memory we want to allow the task to allocate. Obviously, we will reach issues at certain extremes unless we can start to calculate based off of node resources.